<?xml version="1.0"?>
<launch>
     <!-- Base Platform and Drivetrain Constants -->
     <arg name="maxVel" default="2.0"/>
     <arg name="maxTurnRadius" default="0.381"/>
     <arg name="baseWidth" default="0.219"/>
     <arg name="wheelDiameter" default="0.1905"/>
     <arg name="pulsesPerMeter" default="9596"/>

     <!-- Launch configuration settings -->
     <arg name="razor_config_file"       default="$(find razor_imu_9dof)/config/razor.yaml"/>
     <arg name="tf_parent"               default="base_link"/>
     <arg name="imu_topic"               default="/imu/data/filtered" />
     <arg name="odom_topic"              default="dead_reckoning" />
     <arg name="imu_ignore_acc"          default="true" />
     <arg name="imu_remove_gravitational_acceleration" default="true" />


     <node pkg="swanson_base" type="dualclaw_skidsteer_drivetrain_node" name="dualclaw_skidsteer_node" output="screen">
          <param name="serial_device" value="/dev/ttyS0"/>
          <param name="serial_baud" value="115200"/>
          <param name="left_claw_addr" value="128"/>
          <param name="right_claw_addr" value="129"/>
          <param name="base_width" value="$(arg baseWidth)"/>
          <param name="qpps_per_meter" value="$(arg pulsesPerMeter)"/>
          <param name="wheel_diameter" value="$(arg wheelDiameter)"/>
          <param name="max_speed" value="$(arg maxVel)"/>
          <param name="max_turn_radius" value="$(arg maxTurnRadius)"/>
          <param name="update_rate" value="30"/>
     </node>

     <node pkg="razor_imu_9dof" type="imu_node.py" name="imu_node" output="screen">
          <rosparam file="$(arg razor_config_file)" command="load"/>
     </node>

     <node pkg="razor_imu_9dof" type="imu2pose_node.py" name="imu_to_filtered_pose" output="screen" respawn="false" >
          <param name="imu_topic" value="imu/data/filtered" />
          <param name="pose_topic" value="imu/pose/filtered" />
          <param name="verbose" value="true" />
     </node>

     <node pkg="imu_filter_madgwick" type="imu_filter_node" name="imu_filter_madgwick" output="screen" respawn="false" >
          <param name="world_frame" value="enu" />
          <param name="fixed_frame" value="imu_frame" />
          <param name="reverse_tf" value="false" />
          <param name="publish_tf" value="false" />
          <param name="use_mag" value="true" />
          <param name="gain" value="0.2" />
          <param name="use_magnetic_field_msg" value="true" />
          <remap from="imu/data_raw" to="imu/data/raw" />
          <remap from="imu/data" to="imu/data/filtered" />
     </node>

     <node pkg="tf" type="static_transform_publisher" name="tf_imu" args="0 0 0 0 0 0 base_link imu_frame 100"/>
     <!-- <node pkg="tf" type="static_transform_publisher" name="tf_imu_to_cam" args="0 0 0 1.57 0 -1.57 camera_link imu_link 100"/> -->
     <!-- <node pkg="tf" type="static_transform_publisher" name="tf_cam_to_base" args="0 0 0 0 0 0 $(arg tf_parent) camera_link 100"/> -->

     <!-- Odometry fusion (EKF), refer to demo launch file in robot_localization for more info -->
     <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization" clear_params="true" output="screen">
          <rosparam command="load" file="$(find swanson_base)/config/fusion.yaml" />

          <param name="base_link_frame" value="$(arg tf_parent)"/>
          <param name="odom0" value="$(arg odom_topic)"/>
          <param name="imu0" value="$(arg imu_topic)"/>
          <param name="imu0_remove_gravitational_acceleration" value="$(arg imu_remove_gravitational_acceleration)"/>

          <!-- The order of the values is
               x, y, z,
               roll, pitch, yaw,
               vx, vy, vz,
               vroll, vpitch, vyaw,
               ax, ay, az.
          -->
          <rosparam param="odom0_config">
               [true, true, true,
                true, true, true,
                true, true, true,
                false, false, false,
                false, false, false]
          </rosparam>

          <rosparam     if="$(arg imu_ignore_acc)" param="imu0_config">
               [false, false, false,
               true,  true,  true,
               false, false, false,
               true,  true,  true,
               false,  false,  false
               ]
          </rosparam>
          <rosparam unless="$(arg imu_ignore_acc)" param="imu0_config">
               [false, false, false,
               true,  true,  true,
               false, false, false,
               true,  true,  true,
               true,  true,  true
               ]
          </rosparam>

          <!-- The values are ordered as
               x, y, z,
               roll, pitch, yaw,
               vx, vy, vz,
               vroll, vpitch, vyaw,
               ax, ay, az.
          -->
          <rosparam param="process_noise_covariance">
               [0.0005, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
               0,    0.0005, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
               0,    0,    0.006, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
               0,    0,    0,    0.003, 0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
               0,    0,    0,    0,    0.003, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
               0,    0,    0,    0,    0,    0.006, 0,     0,     0,    0,    0,    0,    0,    0,    0,
               0,    0,    0,    0,    0,    0,    0.0025, 0,     0,    0,    0,    0,    0,    0,    0,
               0,    0,    0,    0,    0,    0,    0,     0.0025, 0,    0,    0,    0,    0,    0,    0,
               0,    0,    0,    0,    0,    0,    0,     0,     0.004, 0,    0,    0,    0,    0,    0,
               0,    0,    0,    0,    0,    0,    0,     0,     0,    0.001, 0,    0,    0,    0,    0,
               0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0.001, 0,    0,    0,    0,
               0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0.002, 0,    0,    0,
               0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0.001, 0,    0,
               0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0.001, 0,
               0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0.0015]
          </rosparam>
     </node>



</launch>
